_posts/2025-09-19-minimizing-dependencies.md

```diff
@@ -0,0 +1,122 @@
+---
+title: "The importance of minimizing dependencies in Bitcoin Core"
+permalink: /blog/2025/09/19/minimizing-dependencies/
+layout: post
+author: brink
+name: "brink"
+alt: The importance of minimizing dependencies in Bitcoin Core
+category: "technical"
+description: Relying on external code is a risk. We look at the work in Bitcoin Core to minimize these kind of dependencies.
+---
+
+The security of Bitcoin Core's codebase is rightfully emphasized in discussions
+of the Bitcoin network's resiliency. But Bitcoin Core's code is not the only
+code executed when running Bitcoin Core. Over the years, Bitcoin Core has also
+relied on external, third-party code and libraries. Some of these dependencies
+have been the source of historical vulnerabilities, and all of them carry
+potential risks. Engineers have deliberately and carefully worked to reduce
+dependencies, replacing them with safer, purpose-built code or eliminating them
+entirely.
+
+## Risk of Dependencies
+
+While using pre-existing libraries can speed up development and productivity, in
+Bitcoin Core, securitiy is instead prioritized. Relying on dependencies can
+create security risks in several ways:
+
+- **Security vulnerabilities**: External libraries can contain bugs that expose
+  Bitcoin Core to crashes, remote code execution, or other exploits.
+- **Consensus fragility**: Subtle differences between library versions or
+  platforms can lead to consensus splits. Satoshi’s early use of OpenSSL
+  signature parsing nearly caused just such a problem.
+- **Supply-chain risk**: If an upstream project ships a malicious version,
+  whether by coercion, compromise, or mistake, Bitcoin Core could be exposed to
+  a backdoor.
+- **Build reproducibility**: Reproducible builds, critical for verifiability,
+  are harder to achieve when relying on third-party projects.
+- **Maintenance burden**: Developers must track upstream changes, apply security
+  patches, and ensure cross-platform compatibility.
+
+Every dependency adds to the attack surface. Reducing them increases resilience.
+
+## Removing Dependencies
+
+Over time Bitcoin Core has steadily removed unnecessary and risky dependencies,
+deleting them entirely, or replacing them with Bitcoin-specific, minimal
+implementations. Here are the most significant removals:
+
+- **OpenSSL (replaced with libsecp256k1)**
+  - OpenSSL was used for ECDSA, RNG, and X.509 parsing. Its complexity and
+    vulnerabilities (e.g. Heartbleed) were dangerous.
+  - Replaced by [libsecp256k1][], a purpose-built library, developed by Bitcoin
+    Core contributors.
+  - [Bitcoin Core #6954][bc 6954] (2016).
+
+- **Protobuf / BIP70 Payment Protocol**
+  - Bitcoin Core once supported BIP70, which required Protobuf and OpenSSL’s
+    X.509 certificate handling.
+  - This added attack surface and complexity for a feature rarely used.
+  - Fully removed in [Bitcoin Core #17165][bc 17165] (v0.20.0).
+
+- **Berkeley DB (BDB)**
```

Comment:
Could link to https://github.com/bitcoin/bitcoin/pull/28710. `v30.0`.

Reply:
Added